mitxgraders/comparers/affine_comparer.py

```diff
@@ -0,0 +1,167 @@
+from collections import namedtuple
+import numpy as np
+from voluptuous import Schema, Required, Any, All, Optional, Range
+from mitxgraders.comparers.baseclasses import CorrelatedComparer
+from mitxgraders.helpers.calc.mathfuncs import is_nearly_zero
+
+def get_affine_fit_error(x, y):
+    """
+    Get total error in a linear regression y = ax + b between samples x and y.
+
+    Arguments:
+        x, y: flat numpy array
+
+    Usage
+    =====
+    Zero error in a linear relationship:
+    >>> x = np.array([2, 5, 8])
+    >>> result = get_affine_fit_error(x, 2*x + 1)
+    >>> round(result, 8)
+    0.0
+    """
+    A = np.vstack([x, np.ones(len(x))]).T
+    _, residuals, _, _ = np.linalg.lstsq(A, y, rcond=-1)
+    return np.sqrt(residuals)
+
+def get_proportional_fit_error(x, y):
+    """
+    Get total error in a linear regression y = ax between samples x and y, with
+    zero constant term.
+
+    Arguments:
+        x, y: flat numpy array
+
+    Usage
+    =====
+    Reveals error if relationship is not proportional:
+    >>> x = np.array([2, 5, 8])
+    >>> result = get_proportional_fit_error(x, 2*x + 1)
+    >>> round(result, 8)
+    0.76200076
+
+    Zero error in a proportional relationship:
+    >>> result = get_proportional_fit_error(x, 2*x)
+    >>> round(result, 8)
+    0.0
+    """
+    A = np.vstack(x)
+    _, residuals, _, _ = np.linalg.lstsq(A, y, rcond=-1)
+    return np.sqrt(residuals)
+
+def get_offset_fit_error(x, y):
+    """
+    Get total error in a linear regression y = x + b between samples x and y,
+    with slope term equal to 1.
+
+    Arguments:
+        x, y: flat numpy array
+
+    Usage
+    =====
+    Reveals error if relationship is not constant-offset:
+    >>> x = np.array([2, 5, 8])
+    >>> result = get_offset_fit_error(x, 2*x + 1)
+    >>> round(result, 8)
+    4.24264069
+
+    Zero error in a constant-offset relationship:
+    >>> result = get_offset_fit_error(x, x + 5)
+    >>> round(result, 8)
+    0.0
+    """
+    mean = np.mean(y - x)
+    return np.sqrt(sum(np.square(x + mean - y)))
+
+def get_equals_fit_error(x, y):
+    """
+    Get total error in the difference between two samples.
+    Arguments:
+        x, y: compatible numpy arrays
+    """
+    return sum(np.abs(x-y))
```

Comment:
May I suggest doing this RMS like you do the others?

Reply:
done